File: mpi/exercises/exercise_6/solution.f90
!> @brief Solution to the MPI exercise: 'Switch to non-blocking'.
PROGRAM main
    USE mpi

    IMPLICIT NONE

    INTEGER :: ierror
    INTEGER :: comm_size
    INTEGER :: my_rank
    INTEGER :: value
    INTEGER :: request

    CALL MPI_Init(ierror)

    CALL MPI_Comm_size(MPI_COMM_WORLD, comm_size, ierror)
    IF (comm_size .NE. 2) THEN
        WRITE(*, '(A)') 'This application must be run with 2 MPI processes.'
        CALL MPI_Abort(MPI_COMM_WORLD, -1, ierror)
    END IF

    CALL MPI_Comm_rank(MPI_COMM_WORLD, my_rank, ierror)
    IF (my_rank == 0) THEN
        value = 12345

        ! Launch the non-blocking send
        CALL MPI_Isend(value, 1, MPI_INTEGER, 1, 0, MPI_COMM_WORLD, request, ierror)
        WRITE(*, '(A,I0,A)') '[MPI process ', my_rank ,'] I launched the non-blocking send.'

        ! Wait for the non-blocking send to complete
        CALL MPI_Wait(request, MPI_STATUS_IGNORE, ierror)
        WRITE(*, '(A,I0,A,I0,A)') '[MPI process ', my_rank, '] The wait completed, so I sent value ', value, '.'
    ELSE
        value = 0

        ! Launch the non-blocking receive
        CALL MPI_Irecv(value, 1, MPI_INTEGER, 0, MPI_ANY_TAG, MPI_COMM_WORLD, request, ierror)
        WRITE(*, '(A,I0,A)') '[MPI process ', my_rank ,'] I launched the non-blocking receive.'

        ! Wait for the non-blocking send to complete
        CALL MPI_Wait(request, MPI_STATUS_IGNORE, ierror)
        WRITE(*, '(A,I0,A,I0,A)') '[MPI process ', my_rank, '] The wait completed, so I received value ', value, '.'
    END IF

    CALL MPI_Finalize(ierror)
END PROGRAM main

File: mpi/exercises/exercise_6/solution.c
#include <stdio.h>
#include <mpi.h>

/**
 * @brief Solution to the MPI exercise: "Switch to non-blocking".
 **/
int main(int argc, char* argv[])
{
	MPI_Init(&argc, &argv);

	int comm_size;
	MPI_Comm_size(MPI_COMM_WORLD, &comm_size);
	if(comm_size != 2)
	{
		printf("This application must be run with 2 MPI processes.\n");
		MPI_Abort(MPI_COMM_WORLD, -1);
	}

	int my_rank;
	MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
	if(my_rank == 0)
	{
		int value_sent = 12345;
		MPI_Request request;

		// Launch the non-blocking send
		MPI_Isend(&value_sent, 1, MPI_INT, 1, 0, MPI_COMM_WORLD, &request);
		printf("[MPI process %d] I launched the non-blocking send.\n", my_rank);

		// Wait for the non-blocking send to complete
		MPI_Wait(&request, MPI_STATUS_IGNORE);
		printf("[MPI process %d] The wait completed, so I sent value %d.\n", my_rank, value_sent);
	}
	else
	{
		int value_received = 0;
		MPI_Request request;

		// Launch the non-blocking receive
		MPI_Irecv(&value_received, 1, MPI_INT, 0, MPI_ANY_TAG, MPI_COMM_WORLD, &request);
		printf("[MPI process %d] I launched the non-blocking receive.\n", my_rank);

		// Wait for the non-blocking send to complete
		MPI_Wait(&request, MPI_STATUS_IGNORE);
		printf("[MPI process %d] The wait completed, so I received value %d.\n", my_rank, value_received);
	}

	MPI_Finalize();

	return 0;
}